### DISCUSSION TITLE
How do I run commands in a subshell?

### DISCUSSION THREAD

**Discussion by lieryan** [xonsh/xonsh#4938] [Q&A]
Basically, this is what I want to run (example for zsh and bash):

    $ cd dir1 && cmd1 && ( cd dir2 && cmd2 ) && cmd3

Basically, `cmd1` and `cmd2` should run in `dir1` and `cmd2` should run in `dir2`.

In `bash`, the () syntax creates a subshell which is basically just bash running the subshell commands in a fork; in `zsh`, IIUC, it simulates this mechanism, managing the directory stack and env variables, without actually forking.

Is there a way to do something similar in xonsh? 

What I have tried, this doesn't create a subshell, the `cd` effect isn't reversed when exiting the parentheses:

    $ cd one && ls && ( cd two && ls ) && ls

This and others like it `$()`, `!()`, `$[]`, `![]`, are syntax errors:

    $ cd one && ls && ![ cd two && ls ] && ls
    File "<stdin>", line 1
        cd one && ls && ![ cd two && ls ] && ls
          ^
    SyntaxError: ('code: one',)

Closest I could get is this, but this is clunky as the subshell commands doesn't get syntax highlighted, and it'll probably fail on more complex scenarios with variable substitutions and nested subshells:

    $ cd one && ls && ![ xonsh -c "cd two && ls" ] && ls

One workaround is doing this with a macro, this works ok if all you care about is isolating the cwd changes:

```
def s(cmd):
    return ![ xonsh -c @(cmd) ]

cd one && ls && s!( cd example-project/ && ls ) && ls
```

but it still doesn't handle variable scopes as expected:

```
$ var=1 ; cd one && ls && s!( cd two && ls && print(var) ) && ls
Traceback (most recent call last):
  File "<string>", line 1, in <module>
NameError: name 'var' is not defined
```

**Comment by anki-code** [comment]
If you need this for scripting you can use [`chdir` context manager](https://github.com/xonsh/xonsh/blob/45b12449328a0a76bb95e5b80557342d19c8b03a/tests/test_dirstack.py#L14-L19):
```xsh
from contextlib import contextmanager

@contextmanager
def chdir(adir):
    old_dir = os.getcwd()
    os.chdir(adir)
    yield
    os.chdir(old_dir)
	
# --------------------------------------------

cd /tmp
mkdir -p dir1

pwd
with chdir("./dir1"):
    pwd
pwd

# /tmp
# /tmp/dir1
# /tmp
```

Running commands using the subshell syntax is the new feature for xonsh.

**Reply by lieryan** [reply]
Thanks for replying, @anki-code,

That works when you're writing multi-line scripts, though I usually use subshell syntax on massive one-liners and complex pipelines when in the interactive shell. The return code and piping the stdin/out of the command within the subshell is often needed by the next command.

This is the kind of command that often crops up that doesn't have a straightforward xonsh replacement using a `with` syntax:

```
$ cmd1 | ( cd dir2 && FOO=bar cmd2 ) | cmd3

$ cmd1 <(FOO=bar cmd2 | (cd dir3; cmd3) )
```

That second command would probably require something like [process substitution](https://github.com/xonsh/xonsh/issues/1307) as well though. :thinking:

**Reply by anki-code** [reply]
@lieryan please thumb up #1307

**Reply by lieryan** [reply]
@anki-code done that 😀

**Comment by anki-code** [comment]
hi! I'm going to answer generally because I feel the pressure of posix shells users expectations. The xonsh shell is young and developed by the volunteers without strong release cycles. Current behavior of parser, threading and the logic of how it's working together were constructed for cover pretty minimal needs. So we can't say that the logic is weird we can say that the logic is on the beginner level.

To push the improvement forward:
* create the issue with the clear description of cases, examples, how it works now, how it should work
* become the core developer of xonsh by understanding how it works and do this changes

(I see that my messages has "Maintainer" label but the "maintainer" is the team during a couple of years. Don't treat my comment as a comment from owner.)
